052-kafka-agent-controller-configuration.md

```diff
@@ -0,0 +1,116 @@
+```
+
+This would be recognized by the Kafka Agent as a "running cluster in KRaft mode".
+This way the Kafka Agent would not take care of handling these files for the Kubernetes liveness and readiness probes in ZooKeeper-mode.
+We could think about having a dedicated additional parameter for that, but what we aim with this proposal is anyway a temporary solution and also we are moving away from ZooKeeper, so at some point the first two parameters would be totally removed.
+
+### Exposing KRaft `controller` configuration
+
+During the Kafka start up script, the node configuration is generated and saved on the disk in the `/tmp/strimzi.properties` file.
+As soon as the Kafka Agent runs, it can read this file and load its content in memory in a JSON format.
+
+```json
+{
+    "broker.id": "0",
+    "node.id": "0",
+    "process.roles": "controller",
+    "controller.listener.name": "CONTROL_PLANE-9090",
+    ...
+    ...
+}
+```
+
+Such a configuration is returned through the `/v1/node-configuration` HTTP endpoint.
+Because of this file never changing, even if the node has its configuration dynamically updated, there is no need for the Kafka Agent to watch the file.
+It is enough to load its content on the Kafka Agent startup and keep it as cached in memory.
+The same content will be just returned every time the Kafka Roller is asking for it.
+
+### Kafka Roller usage
+
+The usage of the `controller` node configuration in the Kafka Roller is not a goal of this proposal, but for better understanding it is useful to mention it.
+
+The Kafka Roller queries the Kafka Agent to get the `controller` node configuration in JSON format via HTTPS on the `/v1/node-configuration` endpoint.
+It also has an hardcoded list of the `controller` only configuration properties (which has to be defined taking a look at the Kafka codebase).
+When there is a change in the `Kafka` custom resource, the Kafka Roller checks if some of the `controller` specific parameters are changed, between the desired configuration in the custom resource and the one got from the Kafka Agent.
+If there is any difference, the Kafka Roller just rolls the `controller` node.
+
+## Affected/not affected projects
+
+The main affected project is the Kafka Agent one.
+Also the bash script for running Kafka is affected in order to start the Kafka Agent when KRaft mode is enabled.
+The Kafka Roller is not impacted because the part related to it using the information exposed by the Kafka Agent is not a goal of this proposal.
+
+## Compatibility
+
+This proposal is not going to break any backward compatibility.
+The Kafka Agent will run in KRaft mode as well but not handling the files used by the Kubernetes platform for checking liveness and readiness of the Kafka broker.
+It will still provide broker and log recovery status on the corresponding HTTP endpoint.
+It is going to add one more HTTP endpoint for providing `controller` node configuration.
+
```

Comment:
Is it worth mentioning what will happen if (or what prevents) the operator talking to an older kafka image where the endpoint is not present (and so gets a 404)? I don't think there's anything we can reasonably do in this case. Failing the reconciliation with a suitable error message would be enough.

Reply:
Yeah I just missed that written down, but it was in my notes ... 404 was the idea.